Added support for Material Design via AppCompat library

kanetik opened the pull request:
In order to use Material Design via the AppCompat library, you must extend ActionBarActivity. I've added a new set of AbstractActivity classes that extends this so that developers may use Material Design in plug-ins.

kanetik committed:
Added support for Material Design via AppCompat library

ccjernigan commented on pluginClientSdkLib/build.gradle:
```diff
@@ -8,7 +8,7 @@ dependencies {
     // https://code.google.com/p/android/issues/detail?id=68316
     compile group:'com.android.support', name:'support-v4', version:']19,)'
-
+    compile group:'com.android.support', name:'appcompat-v7', version:'[21,)'
```
Should the version matcher for support-v4 and appcompat-v7 be the same?

Both are bundled in the local maven repo installed by the Android SDK manager, so it would be unlikely for version 20 of support-v4 to be used with version 21 of app-compat in practice.  But I still think the two libraries probably should be in sync

Also the ] in front of support-v4's version matcher was an artifact of the Lollipop preview release causing some Gradle grief.  '[21,)' should be OK for both support-v4 and appcompat-v7.

ccjernigan replied:
The maven portion of the build script needs to be modified to filter out appcompat-v7 for clients that don't wish to use it.  We're already doing this for support-v4.  Search for "pom.whenConfigured" in this file to find it.

ccjernigan commented:
The build is failing due to lint errors.  `./gradlew check` is the command to run lint locally.  That should then make it easy to see what lint is flagging.

kanetik committed:
Merge branch 'master' of https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale

kanetik commented:
I'm working through the lint errors, and this one seems to be a problem:
when using appCompat, the build fails as long as the menu uses
"android:showAsAction". You need to instead use "app:showAsAction", adding
the app namespace.

The challenge there, of course, is that if we make appCompat (and support)
optional in the gradle file, then wouldn't that mean "app:showAsAction"
will cause the build to fail if the user is not including the appCompat
library?

I'm presuming you are not interested in requiring the appCompat & support
library, so I'm going to make that setting happen programmatically, but
before I did that, I wanted to check if you knew of another solution to the
problem.

Thanks!
J

On Sat, Feb 21, 2015 at 11:13 PM, Carter Jernigan notifications@github.com
wrote:

> The build is failing due to lint errors. ./gradlew check is the command
> to run lint locally. That should then make it easy to see what lint is
> flagging.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#issuecomment-75418750
> .

kanetik committed:
Lint corrections
PR Review Corrections
Proguard config change for appCompat
No need for API Level checks for appCompat methods

kanetik committed:
Class name correction

kanetik committed:
missed one

kanetik commented:
Following up on my own email - I've committed (and will shortly PR) a
change that is not quite as involved as I had previously discussed. I've
fixed all the issues that the Lint report showed, and that you mentioned,
and a few that I found on my own. I believe that by simply configuring the
lint check to ignore the android vs app namespace property, the worst that
will happen is that, when appCompat is being used, it will ignore the menu
setting - but I think that's okay, as I believe anyone creating a Material
Design plugin should be considering very carefully how the menu is
displayed anyhow.

I'll create a PR, and let you review it and see if there are any other
changes you'd recommend.

Thanks!
J

On Sun, Feb 22, 2015 at 12:22 PM, Jeremy Kane jkane001@gmail.com wrote:

> I'm working through the lint errors, and this one seems to be a problem:
> when using appCompat, the build fails as long as the menu uses
> "android:showAsAction". You need to instead use "app:showAsAction", adding
> the app namespace.
> 
> The challenge there, of course, is that if we make appCompat (and support)
> optional in the gradle file, then wouldn't that mean "app:showAsAction"
> will cause the build to fail if the user is not including the appCompat
> library?
> 
> I'm presuming you are not interested in requiring the appCompat & support
> library, so I'm going to make that setting happen programmatically, but
> before I did that, I wanted to check if you knew of another solution to the
> problem.
> 
> Thanks!
> J
> 
> On Sat, Feb 21, 2015 at 11:13 PM, Carter Jernigan <
> notifications@github.com> wrote:
> 
> > The build is failing due to lint errors. ./gradlew check is the command
> > to run lint locally. That should then make it easy to see what lint is
> > flagging.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#issuecomment-75418750
> > .

kanetik commented:
Fixed a few issues and refactored a little

kanetik committed:
Updated example in javadoc attribute
Include appcompat and support in testbuild for Travis CI

kanetik commented:
Ok, Travis CI and I are not friends. Looking into the root cause of this latest build failure, it seems like perhaps a Travis CI bug or configuration issue with the support library, but I'm not sure. I've tried one of the fixes mentioned in my searching, but that didn't affect anything. Thoughts?

ccjernigan commented on pluginClientSdkLib/build.gradle:
```diff
@@ -17,6 +17,8 @@ dependencies {
     compile group:'com.twofortyfouram', name:'android-spackle', version:'[1.0.4,2.0['
 
     androidTestCompile group:'com.twofortyfouram', name:'android-test', version:'[1.0.4,2.0['
+    testCompile fileTree(dir: "$rootProject.projectDir/build/intermediates/exploded-aar/com.android.support/support-v4/", include: "**/*.jar")
```
What problem are these lines intended to fix?

We don't have any unit tests introduced in the Android Gradle plugin 1.1.0, so testCompile shouldn't be necessary as far as I know.

ccjernigan replied:
This question is still open: what is the testCompile option here supposed to do?

ccjernigan commented on pluginClientSdkLib/build.gradle:
```diff
@@ -78,6 +80,8 @@ uploadArchives {
             // Exclude support.v4, making it an optional dependency
             pom.whenConfigured { pom -> pom.dependencies.removeAll {dep -> dep.groupId == 'com.android.support' && dep.artifactId == 'support-v4' } }
+            // Exclude appcompat-v7, making it an optional dependency
+            pom.whenConfigured { pom -> pom.dependencies.removeAll {dep -> dep.groupId == 'com.android.support' && dep.artifactId == 'appcompat-v7' } }
```
I think this is broken.  The way I read this: we're assigning the pom.whenConfigured variable to a closure.  Then we re-assign to a different closure right after, dereferencing the first closure.

I think the correct solution involves a single closure with `dep.groupId == 'com.android.support' && (dep.artifactId == 'support-v4' || dep.artifactId == 'appcompat-v7')`

One way to test whether this worked as intended is to run `./gradlew installArchives` and then looking at the local maven repo (on my Mac, under ~/m2/).  The pom file nested in the maven repo lists the required dependencies, and we'd be looking for the absence of support-v4 and appcompat-v7.

kanetik replied:
I've run gradlew installArchives with the changes I made commented out, and
I am not seeing any evidence of appCompat in the pom file
(.m2\repository\com\twofortyfouram\android-plugin-client-sdk-for-locale\1.0.0\android-plugin-client-sdk-for-locale-1.0.0.pom)
as I would expect to:

<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="
http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.twofortyfouram</groupId>
  <artifactId>android-plugin-client-sdk-for-locale</artifactId>
  <version>1.0.0</version>
  <packaging>aar</packaging>
</project>

I could not find good documentation around pom.whenConfigured, but I read
each line as a directive, not a variable, so I expected that it would run
one after the other, but I could be wrong (and likely am).

On Tue, Feb 24, 2015 at 10:28 AM, Carter Jernigan notifications@github.com
wrote:

> In pluginClientSdkLib/build.gradle
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#discussion_r25260285
> :
> 
> > @@ -78,6 +80,8 @@ uploadArchives {
> >              }
> >              // Exclude support.v4, making it an optional dependency
> >              pom.whenConfigured { pom -> pom.dependencies.removeAll {dep -> dep.groupId == 'com.android.support' && dep.artifactId == 'support-v4' } }
> > -            // Exclude appcompat-v7, making it an optional dependency
> > -            pom.whenConfigured { pom -> pom.dependencies.removeAll {dep -> dep.groupId == 'com.android.support' && dep.artifactId == 'appcompat-v7' } }
> 
> I think this is broken. The way I read this: we're assigning the
> pom.whenConfigured variable to a closure. Then we re-assign to a different
> closure right after, dereferencing the first closure.
> 
> I think the correct solution involves a single closure with dep.groupId
> == 'com.android.support' && (dep.artifactId == 'support-v4' ||
> dep.artifactId == 'appcompat-v7')
> 
> One way to test whether this worked as intended is to run ./gradlew
> installArchives and then looking at the local maven repo (on my Mac,
> under ~/m2/). The pom file nested in the maven repo lists the required
> dependencies, and we'd be looking for the absence of support-v4 and
> appcompat-v7.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1/files#r25260285
> .

ccjernigan commented:
Did you take a look at the Travis log?  It appears that JavaDoc generation is causing the build failure.  I'm not clear why specifically that is, so I'll have to look more deeply at it later today

Regarding the menu items: I need to look into a couple of things to figure out what makes the most sense because I've actually not used the appcompat library until now.

Regarding the overall approach to Material design support for plug-ins, I think we need to give some consideration to what makes sense.  The UI for plug-ins was designed for Android 1.0, then we've migrated through a half-dozen Android UI updates.  Some of the things we did over the last 7 years may not make sense anymore, such as breadcrumbs or showing the host app's icon, given how drastically different Material design is.  What are your thoughts?

I suppose I'm considering two approaches to move forward:
- Spend more time figuring out what the AbstractLocaleActionBarPluginActivity from a UI design perspective.  So we might need to have a little more back and forth on that activity to get it right.
- Reduce the scope of this pull request by only including AbstractActionBarPluginActivity (and remove AbstractLocaleActionBarPluginActivity for now).  That would expedite completion of this pull request, by limiting its scope to what we know for sure.  Then we could create a second pull request to deal with adding AbstractLocaleActionBarPluginActivity and the associated UI discussion.

Thoughts?

ccjernigan commented:
Also just FYI: the complete task that we're running on Travis is found near the bottom of the .travis.yml file in the root of the repo.  That shows how we're running the JavaDoc task.

ccjernigan commented:
And one more code review comment: we'll need to increment the MINOR version number of the SDK in the gradle.properties file since we're adding new features but they're backwards compatible.  We're using semantic versioning: semver.org

So the version number would end up being 1.1.0

kanetik commented:
I'm at work (full-time job, where the plug-ins are pretty much just my
hobby) at the moment, so I'll have to get back to you on all this tonight.
Quick answer about the Travis CI build failing - yes, I saw that it was the
JavaDoc step that was failing, and the failure is that it cannot find a
class that is in the appCompat library. One of the possible fixes I'd found
in researching that was to add the androidTestCompile, but it didn't work,
and it makes sense that it wouldn't, since the build isn't a test.  In any
event, seems that Travis isn't loading the appCompat library, and some of
the other research I'd done basically says that Travis is using an old
version of BuildTools and can be manually configured to use the latest
version, or that it might be a Travis bug, but I'm not sure there's much
you can do about either of those - not sure how much control you have over
that Travis config.

I'll go through the rest of your comments and respond later, there's a lot
of good stuff there!

Thanks
J

On Tue Feb 24 2015 at 10:41:15 AM Carter Jernigan notifications@github.com
wrote:

> And one more code review comment: we'll need to increment the MINOR
> version number of the SDK in the gradle.properties file since we're adding
> new features but they're backwards compatible. We're using semantic
> versioning: semver.org
> 
> So the version number would end up being 1.1.0
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#issuecomment-75779934
> .

kanetik commented:
I do think there is a pretty massive push from Google (the likes I've not
seen before) to move toward Material Design. Lots of apps are making the
switch (including my employer's apps, which I've been working on for the
past 3 months), so I think there's plenty of inertia behind this, unlike
what was behind Ice Cream Sandwich or Jelly Bean.

That said, I can understand why you wouldn't want to force Material Design
on users of the SDK, but I wonder if there can be a line drawn in the sand,
where you say that devs who want to stick with the old UI conventions can
continue to do so, using the existing library/SDK; users of the new SDK &
gradle dependency would require support and appCompat libraries.

Not sure you can (or want to) make that a requirement, but it would
obviously reduce the complexity of these sorts of things.

In terms of the 2 options you laid out, I think I'd prefer the first
option. I can use a local version of the dependency while I develop my
Material/Android Studio revisions of my plugins (I'm moving slowly in that
regard, since I have client work that's been taking most of my time), so
I'm not in a massive rush to have the official appCompat-friendly solution
just yet - so I'm perfectly fine going back and forth.

Thanks!
J

On Tue, Feb 24, 2015 at 10:38 AM, Carter Jernigan notifications@github.com
wrote:

> Did you take a look at the Travis log? It appears that JavaDoc generation
> is causing the build failure. I'm not clear why specifically that is, so
> I'll have to look more deeply at it later today
> 
> Regarding the menu items: I need to look into a couple of things to figure
> out what makes the most sense because I've actually not used the appcompat
> library until now.
> 
> Regarding the overall approach to Material design support for plug-ins, I
> think we need to give some consideration to what makes sense. The UI for
> plug-ins was designed for Android 1.0, then we've migrated through a
> half-dozen Android UI updates. Some of the things we did over the last 7
> years may not make sense anymore, such as breadcrumbs or showing the host
> app's icon, given how drastically different Material design is. What are
> your thoughts?
> 
> I suppose I'm considering two approaches to move forward:
> - Spend more time figuring out what the
>   AbstractLocaleActionBarPluginActivity from a UI design perspective. So we
>   might need to have a little more back and forth on that activity to get it
>   right.
> - Reduce the scope of this pull request by only including
>   AbstractActionBarPluginActivity (and remove
>   AbstractLocaleActionBarPluginActivity for now). That would expedite
>   completion of this pull request, by limiting its scope to what we know for
>   sure. Then we could create a second pull request to deal with adding
>   AbstractLocaleActionBarPluginActivity and the associated UI discussion.
> 
> Thoughts?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#issuecomment-75779351
> .

kanetik committed:
Cleanup

kanetik committed:
Travis Fix?

kanetik committed:
Version Bump

kanetik committed:
Include Android Maven Repository

kanetik committed:
Attempting to resolve Javadoc issues with Support Library

kanetik committed:
Feeling pretty good about this one...

kanetik committed:
Stupid Typo

kanetik committed:
Tired of seeing this lint warning

kanetik committed:
Now that I can get the GenerateReleaseJavadoc to run locally, I think I have gotten the last issue resolved. Fingers crossed.

ccjernigan commented on pluginClientSdkLib/build.gradle:
```diff
@@ -57,6 +58,14 @@ android.libraryVariants.all { variant ->
     }
 }
 
+if (JavaVersion.current().isJava8Compatible()) {
```
What problem were you running into that needed this option to be set?  When building locally, I don't see any warnings with this block removed.

kanetik replied:
I think both of those can be removed, I'll update later.

On Sun, Mar 15, 2015, 4:16 PM Carter Jernigan notifications@github.com
wrote:

> In pluginClientSdkLib/build.gradle
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#discussion_r26451749
> :
> 
> > @@ -57,6 +58,14 @@ android.libraryVariants.all { variant ->
> >      }
> >  }
> > 
> > +if (JavaVersion.current().isJava8Compatible()) {
> 
> What problem were you running into that needed this option to be set? When
> building locally, I don't see any warnings with this block removed.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1/files#r26451749
> .

ccjernigan commented on pluginClientSdkLib/src/main/java/com/twofortyfouram/locale/sdk/client/receiver/AbstractAsyncReceiver.java:
```diff
@@ -62,6 +62,7 @@
                 new Pair<PendingResult, AsyncCallback>(pendingResult, callback)));
     }
 
+    @SuppressWarnings("unchecked")
```
I usually try limit the scope of a suppress annotation as much as possible, so that future warnings don't get accidentally suppressed.  What about a small helper method to perform the cast, and only suppress warnings for that helper method?  For example:

```
    @NonNull
    @SuppressWarnings("unchecked")
    private static Pair<PendingResult, AsyncCallback> castObj(@NonNull final Object o) {
        return (Pair<PendingResult, AsyncCallback>) o;
    }
```

Then the Handler could call:

```
    final Pair<PendingResult, AsyncCallback> pair = castObj(msg.obj);
```

kanetik replied:
That feels a bit over-engineered to me - if you don't mind the lint
warning, I will just delete the @SuppressWarnings("unchecked").

J

On Sun, Mar 15, 2015 at 4:42 PM Carter Jernigan notifications@github.com
wrote:

> In
> pluginClientSdkLib/src/main/java/com/twofortyfouram/locale/sdk/client/receiver/AbstractAsyncReceiver.java
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1#discussion_r26452085
> :
> 
> > @@ -62,6 +62,7 @@
> >                  new Pair<PendingResult, AsyncCallback>(pendingResult, callback)));
> >      }
> > -    @SuppressWarnings("unchecked")
> 
> I usually try limit the scope of a suppress annotation as much as
> possible, so that future warnings don't get accidentally suppressed. What
> about a small helper method to perform the cast, and only suppress warnings
> for that helper method? For example:
> 
> ```
> @NonNull
> @SuppressWarnings("unchecked")
> private static Pair<PendingResult, AsyncCallback> castObj(@NonNull final Object o) {
>     return (Pair<PendingResult, AsyncCallback>) o;
> }
> ```
> 
> Then the Handler could call:
> 
> ```
> final Pair<PendingResult, AsyncCallback> pair = castObj(msg.obj);
> ```
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/twofortyfouram/android-plugin-client-sdk-for-locale/pull/1/files#r26452085
> .

ccjernigan replied:
That works too.  I don't think we've hit over-engineering until there's at least unit test for that private helper method :-)

kanetik committed:
Remove unnecessary code

kanetik committed:
This one was needed - but added a comment explaining it.

kanetik committed:
More accurate comment

kanetik commented:
Don't know if you saw the latest commits - I removed all the unnecessary code, but the isJava8Compatible code block in the gradle file is needed to prevent lint errors in the javadoc job to cause it to fail, since the lint errors are not legit (it doesn't like the formatting of the "code" in the javadoc, I think). This is a new "feature" of Java8, and Travis CI seems to be using Java 8 now, I guess?

See http://blog.joda.org/2014/02/turning-off-doclint-in-jdk-8-javadoc.html

ccjernigan commented:
After additional design discussion, I've made a few changes to this pull request, merged those changes to our master branch, and released a new 1.1.0 version of the SDK with these changes.  Thank you for the contributions!
